// ngx-videogular/esm5/lib/streaming/vg-streaming.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VgDASH } from './vg-dash/vg-dash';
import { VgHLS } from './vg-hls/vg-hls';
import * as ɵngcc0 from '@angular/core';
export * from './vg-dash/vg-dash';
export * from './vg-hls/vg-hls';
var VgStreamingModule = /** @class */ (function () {
    function VgStreamingModule() {
    }
VgStreamingModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: VgStreamingModule });
VgStreamingModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function VgStreamingModule_Factory(t) { return new (t || VgStreamingModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(VgStreamingModule, { declarations: function () { return [VgDASH, VgHLS]; }, imports: function () { return [CommonModule]; }, exports: function () { return [VgDASH, VgHLS]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(VgStreamingModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [
                    VgDASH, VgHLS
                ],
                exports: [
                    VgDASH, VgHLS
                ]
            }]
    }], function () { return []; }, null); })();
    return VgStreamingModule;
}());
export { VgStreamingModule };
