// @ng-bootstrap/ng-bootstrap/esm5/buttons/label.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var NgbButtonLabel = /** @class */ (function () {
    function NgbButtonLabel() {
    }
NgbButtonLabel.ɵfac = function NgbButtonLabel_Factory(t) { return new (t || NgbButtonLabel)(); };
NgbButtonLabel.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NgbButtonLabel, selectors: [["", "ngbButtonLabel", ""]], hostVars: 8, hostBindings: function NgbButtonLabel_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("btn", true)("active", ctx.active)("disabled", ctx.disabled)("focus", ctx.focused);
    } } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgbButtonLabel, [{
        type: Directive,
        args: [{
                selector: '[ngbButtonLabel]',
                host: { '[class.btn]': 'true', '[class.active]': 'active', '[class.disabled]': 'disabled', '[class.focus]': 'focused' }
            }]
    }], function () { return []; }, null); })();
    return NgbButtonLabel;
}());
export { NgbButtonLabel };
if (false) {
    /** @type {?} */
    NgbButtonLabel.prototype.active;
    /** @type {?} */
    NgbButtonLabel.prototype.disabled;
    /** @type {?} */
    NgbButtonLabel.prototype.focused;
}
