// ngx-videogular/esm5/lib/ima-ads/ima-ads.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VgImaAds } from './vg-ima-ads';
import * as ɵngcc0 from '@angular/core';
var VgImaAdsModule = /** @class */ (function () {
    function VgImaAdsModule() {
    }
VgImaAdsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: VgImaAdsModule });
VgImaAdsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function VgImaAdsModule_Factory(t) { return new (t || VgImaAdsModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(VgImaAdsModule, { declarations: function () { return [VgImaAds]; }, imports: function () { return [CommonModule]; }, exports: function () { return [VgImaAds]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(VgImaAdsModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [
                    VgImaAds
                ],
                exports: [
                    VgImaAds
                ]
            }]
    }], function () { return []; }, null); })();
    return VgImaAdsModule;
}());
export { VgImaAdsModule };
