// ngx-videogular/esm5/lib/core/vg-core.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { VgControlsHidden } from './services/vg-controls-hidden';
import { VgFullscreenAPI } from './services/vg-fullscreen-api';
import { VgUtils } from './services/vg-utils';
import { VgAPI } from './services/vg-api';
import { VgMedia } from './vg-media/vg-media';
import { VgCuePoints } from './vg-cue-points/vg-cue-points';
import { VgPlayer } from './vg-player/vg-player';
import { VgStates } from './states/vg-states';
import { VgEvents } from './events/vg-events';
import * as ɵngcc0 from '@angular/core';
export * from './vg-media/vg-media-element';
export * from './services/vg-api';
export * from './services/vg-fullscreen-api';
export * from './services/vg-utils';
export * from './services/vg-controls-hidden';
export * from './events/vg-events';
export * from './states/vg-states';
var VgCoreModule = /** @class */ (function () {
    function VgCoreModule() {
    }
VgCoreModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: VgCoreModule });
VgCoreModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function VgCoreModule_Factory(t) { return new (t || VgCoreModule)(); }, providers: [VgAPI, VgFullscreenAPI, VgUtils, VgControlsHidden, VgStates, VgEvents] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(VgCoreModule, { declarations: function () { return [VgMedia, VgCuePoints, VgPlayer]; }, exports: function () { return [VgMedia, VgCuePoints, VgPlayer]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(VgCoreModule, [{
        type: NgModule,
        args: [{
                declarations: [VgMedia, VgCuePoints, VgPlayer],
                exports: [VgMedia, VgCuePoints, VgPlayer],
                providers: [VgAPI, VgFullscreenAPI, VgUtils, VgControlsHidden, VgStates, VgEvents]
            }]
    }], function () { return []; }, null); })();
    return VgCoreModule;
}());
export { VgCoreModule };
