// ngx-videogular/esm5/lib/overlay-play/vg-overlay-play.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VgOverlayPlay } from './vg-overlay-play';
import * as ɵngcc0 from '@angular/core';
var VgOverlayPlayModule = /** @class */ (function () {
    function VgOverlayPlayModule() {
    }
VgOverlayPlayModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: VgOverlayPlayModule });
VgOverlayPlayModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function VgOverlayPlayModule_Factory(t) { return new (t || VgOverlayPlayModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(VgOverlayPlayModule, { declarations: function () { return [VgOverlayPlay]; }, imports: function () { return [CommonModule]; }, exports: function () { return [VgOverlayPlay]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(VgOverlayPlayModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [
                    VgOverlayPlay
                ],
                exports: [
                    VgOverlayPlay
                ]
            }]
    }], function () { return []; }, null); })();
    return VgOverlayPlayModule;
}());
export { VgOverlayPlayModule };
