// ngx-videogular/esm2015/lib/buffering/vg-buffering.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VgBuffering } from './vg-buffering';
import * as ɵngcc0 from '@angular/core';
let VgBufferingModule = class VgBufferingModule {
};
VgBufferingModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: VgBufferingModule });
VgBufferingModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function VgBufferingModule_Factory(t) { return new (t || VgBufferingModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(VgBufferingModule, { declarations: function () { return [VgBuffering]; }, imports: function () { return [CommonModule]; }, exports: function () { return [VgBuffering]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(VgBufferingModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [
                    VgBuffering
                ],
                exports: [
                    VgBuffering
                ]
            }]
    }], null, null); })();
export { VgBufferingModule };
